tsdb/index/tsi1/cache.go

```diff
@@ -0,0 +1,175 @@
+func (c *TagValueSeriesIDCache) Get(name, key, value []byte) *tsdb.SeriesIDSet {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.get(name, key, value)
+}
+
+func (c *TagValueSeriesIDCache) get(name, key, value []byte) *tsdb.SeriesIDSet {
+	if mmap, ok := c.cache[string(name)]; ok {
+		if tkmap, ok := mmap[string(key)]; ok {
+			if ele, ok := tkmap[string(value)]; ok {
+				c.evictor.MoveToFront(ele) // This now becomes most recently used.
+				return ele.Value.(*seriesIDCacheElement).SeriesIDSet
+			}
+		}
+	}
+	return nil
+}
+
+// exists returns true if the an item exists for the tuple {name, key, value}.
+func (c *TagValueSeriesIDCache) exists(name, key, value []byte) bool {
+	if mmap, ok := c.cache[string(name)]; ok {
+		if tkmap, ok := mmap[string(key)]; ok {
+			_, ok := tkmap[string(value)]
+			return ok
+		}
+	}
+	return false
+}
+
+// addToSet adds x to the SeriesIDSet associated with the tuple {name, key, value}
+// if it exists. This method takes a lock on the underlying SeriesIDSet.
+//
+// NB this does not count as an access on the set—therefore the set is not promoted
+// within the LRU cache.
+func (c *TagValueSeriesIDCache) addToSet(name, key, value []byte, x uint64) {
+	if mmap, ok := c.cache[string(name)]; ok {
+		if tkmap, ok := mmap[string(key)]; ok {
+			if ele, ok := tkmap[string(value)]; ok {
+				ss := ele.Value.(*seriesIDCacheElement).SeriesIDSet
+				if ss == nil {
+					ele.Value.(*seriesIDCacheElement).SeriesIDSet = tsdb.NewSeriesIDSet(x)
+					return
+				}
+				ele.Value.(*seriesIDCacheElement).SeriesIDSet.Add(x)
+			}
+		}
+	}
+}
+
+// measurementContainsSets returns true if there are sets cached for the provided measurement.
+func (c *TagValueSeriesIDCache) measurementContainsSets(name []byte) bool {
+	_, ok := c.cache[string(name)]
+	return ok
+}
+
+// Put adds the SeriesIDSet to the cache under the tuple {name, key, value}. If
+// the cache is at its limit, then the least recently used item is evicted.
+func (c *TagValueSeriesIDCache) Put(name, key, value []byte, ss *tsdb.SeriesIDSet) {
+	c.mu.Lock()
+	// Check under the write lock if the relevant item is now in the cache.
+	if c.exists(name, key, value) {
+		c.mu.Unlock()
+		return
+	}
+	defer c.mu.Unlock()
+
+	// Create list item, and add to the front of the eviction list.
+	listElement := c.evictor.PushFront(&seriesIDCacheElement{
+		name:        name,
+		key:         key,
+		value:       value,
+		SeriesIDSet: ss,
+	})
+
+	// Add the listElement to the set of items.
+	if mmap, ok := c.cache[string(name)]; ok {
+		if tkmap, ok := mmap[string(key)]; ok {
+			if _, ok := tkmap[string(value)]; ok {
+				// FIXME(edd): only here whilst testing.
+				fmt.Println("existence of cache item breaks invariant")
+				goto EVICT
+			}
+
+			// Add the set to the map
+			tkmap[string(value)] = listElement
+			goto EVICT
+		}
+
+		// No series set map for the tag key - first tag value for the tag key.
+		mmap[string(key)] = map[string]*list.Element{string(value): listElement}
```

Comment:
I don't know how often these lines happen vs the `tkmap[string(value)] = listElement` above, but i think doing it like this causes an extra allocation on `string(value)`. Maybe the compiler is smart enough, though?

Reply:
I'll note that for the final PR but I don't think it's a biggie for this experiment.